rsvg-convert.1
.\" -*- fill-column:100 -*-
.TH rsvg-convert 1
.SH NAME
rsvg-convert \- Render SVG documents to PNG images, or convert them to PDF or PS.
.SH SYNOPSIS
Convert an SVG to PNG at its "natural size" and write it to standard output:
.P
.RS
.B rsvg-convert
.I input.svg
.B >
.I output.png
.RE
.P
Specify an output filename; the input filename must be the last argument:
.P
.RS
.B rsvg-convert
.BI --output= output.png
.I input.svg
.RE
.P
Configure dots-per-inch (DPI), default is 96:
.P
.RS
.B rsvg-convert
.BI --dpi-x= 300
.BI --dpi-y= 300
.I input.svg
.B >
.I output.png
.RE
.P
Render an SVG at a specific pixel size, scaled proportionally:
.P
.RS
.B rsvg-convert
.BI --width= 1024
.BI --height= 768
.B --keep-aspect-ratio
.I input.svg
.B >
.I output.png
.RE
.P

.SH DESCRIPTION
.BR rsvg-convert
renders SVG documents into PNG raster images, or converts them to PDF or PS as vector objects.
By default
.BR rsvg-convert
will render an SVG document to a raster PNG image and write it to standard output:
.P
.RS
.B rsvg-convert
.I input.svg
.B >
.I output.png
.RE
.P
To select another format, use the
.B --format
option:
.P
.RS
.B rsvg-convert --format=pdf
.I input.svg
.B >
.I output.pdf
.RE
.P
You can use
.BR rsvg-convert
as part of a pipeline; without an argument for the input filename it will read the document from standard input:
.P
.RS
.B cat
.I input.svg
.B |
.B rsvg-convert
.B >
.I output.png
.RE
.P
.SS SPECIFYING THE RENDERED SIZE
You can use the
.B --width
and
.B --height
options to specify the size of the output image.  Most of the time you should specify
.B --keep-aspect-ratio
to scale the image proportionally; for compatibility with old versions this is not the default.
.P
.RS
.B rsvg-convert
.BI --width= 100
.BI --height= 200
.B --keep-aspect-ratio
.I input.svg
.B >
.I output.png
.RE
.P
You can also specify dimensions as CSS lengths, for example
.B 10px
or \"
.BR 8.5in .
The unit specifiers supported are as follows:
.RS
.TS
tab (@);
l lx.
px@T{
pixels (the unit specifier can be omitted)
T}
in@T{
inches
T}
cm@T{
centimeters
T}
mm@T{
millimeters
T}
pt@T{
points, 1/72 inch
T}
pc@T{
picas, 1/6 inch
T}
.TE
.RE
.P
The following will create a 600*900 pixel PNG, or 2*3 inches at 300 dots-per-inch:
.P
.RS
.B rsvg-convert
.BI --width= 2in
.BI --height= 3in
.B --keep-aspect-ratio
.BI --dpi-x= 300
.BI --dpi-y= 300
.I input.svg
.B >
.I output.png
.RE
.P
This will scale an SVG document to fit in an A4 page and convert it to PDF:
.P
.RS
.B rsvg-convert
.BI --format= pdf
.BI --width= 210mm
.BI --height= 297mm
.B --keep-aspect-ratio
.I input.svg
.B >
.I output.pdf
.RE
.P

.SS SPECIFYING A PAGE SIZE
By default the size of the output comes from the rendered size, which can be specified with the
.B --width
and
.B --height
options, but you can specify a page size independently of the rendered size with
.B --page-width
and
.BR --page-height ,
together with
.B --top
and
.B --left
to control the position of the rendered image within the page.

This will create a PDF with a landscape A4 page, by scaling an SVG document to 10*10\ cm, and
placing it with its top-left corner 5\ cm away from the top and 8\ cm from the left of the page:
.P
.RS
.B rsvg-convert
.BI --format= pdf
.BI --page-width= 297mm
.BI --page-height= 210mm
.BI --width= 10cm
.BI --height= 10cm
.B --keep-aspect-ratio
.BI --top= 5cm
.BI --left= 8cm
.I input.svg
.B >
.I output.pdf
.RE
.P

.SS SPECIFYING A SCALE FACTOR INSTEAD OF A RENDERED SIZE
The
.B --zoom
option lets you scale the natural size of an SVG document.  For example, if
.I input.svg
is a document with a declared size of 100*200\ pixels, then the following command will render it at 250*500\ pixels (zoom\ 2.5):
.P
.RS
.B rsvg-convert
.BI --zoom=2.5
.I input.svg
.B >
.I output.png
.RE
.P
You can limit the maximum scaled size by specifying the
.B --width
and
.B --height
options together with
.BR --zoom.
Here, the image will be scaled 10x, but limited to 1000*1000\ pixels at the most:
.P
.RS
.B rsvg-convert
.BI --zoom=10
.BI --width=1000
.BI --height=1000
.I input.svg
.B >
.I output.png
.RE
.P
If you need different scale factors for the horizontal and vertical dimensions, use the
.B --x-zoom
and
.B --y-zoom
options instead of
.BR --zoom.

.SS CREATING A MULTI-PAGE DOCUMENT
The "pdf", "ps", and "eps" output formats support multiple pages. These can be created by
combining multiple input SVG files. For example, this PDF file will have three pages:
.P
.RS
.B rsvg-convert
.BI --format= pdf
.I pg1.svg
.I pg2.svg
.I pg3.svg
.B >
.I out.pdf
.RE
.P
The size of each page will be computed, separately, as described in the
.B "DEFAULT OUTPUT SIZE"
section. This may result in a PDF being produced with differently-sized pages.
If you need to produce a PDF with all pages set to exactly the same size,
use the
.B --page-width
and
.B --page-height
options.
.P
For example, the following command creates a three-page PDF out of three SVG documents.  All the
pages are portrait US Letter, and each SVG is scaled to fit so that there is a 1in margin around
each page:
.P
.RS
.B rsvg-convert
.BI --format= pdf
.BI --page-width= 8.5in
.BI --page-height= 11in
.BI --width= 6.5in
.BI --height= 9in
.B --keep-aspect-ratio
.BI --top= 1in
.BI --left= 1in
.I pg1.svg
.I pg2.svg
.I pg3.svg
.B >
.I out.pdf
.RE
.P
.SS CONVERSION OF PIXELS BASED ON THE DOTS-PER-INCH
.B rsvg-convert
uses the
.B --dpi-x
and
.B --dpi-y
options to configure the dots-per-inch (DPI) by which pixels will be converted to/from physical units like inches or centimeters.  The default for both options is 96\ DPI.

Consider this example SVG, which is nominally declared to be 2*3 inches in size:
.P
.in +4n
.EX
<svg xmlns="http://www.w3.org/2000/svg" width="2in" height="3in">
  <!-- graphical objects here -->
</svg>
.EE
.in
.P
The following commands create PNGs of different sizes for the example SVG above:
.P
.RS
.B rsvg-convert
.I two-by-three.svg
.B >
.I output.png
#### creates a 192*288\ pixel PNG
.P
.B rsvg-convert
.BI --dpi-x= 300
.BI --dpi-y= 300
.I two-by-three.svg
.B >
.I output.png
#### creates a 600*900\ pixel PNG
.RE
.P

Note that the final pixel dimensions are rounded up to the nearest pixel, to avoid
clipping off the right/bottom edges.  In the following example,
.B rsvg-convert
will generate a PNG 300x300 pixels in size:
.P
.RS
.B rsvg-convert
.BI --width= 299.5
.BI --height= 299.4
.I input.svg
.B >
.I output.png
#### outputs 300x300 pixel PNG with a fractionally-scaled image
.RE
.P
If you specify dimensions in physical units, they will be multiplied by the dots-per-inch (DPI) value to obtain
dimensions in pixels.  For example, this will generate a 96x96 pixel PNG, since it is 1x1 inch at the default 96\ DPI:
.P
.RS
.B rsvg-convert
.BI --width= 1in
.BI --height= 1in
.I input.svg
.B >
.I output.png
#### outputs 96x96 pixel PNG
.RE
.P
Correspondingly, this will generate a 300x300 pixel PNG, since it is 1x1 inch at 300 DPI:
.P
.RS
.B rsvg-convert
.BI --width= 1in
.BI --height= 1in
.BI --dpi-x= 300
.BI --dpi-y= 300
.I input.svg
.B >
.I output.png
#### outputs 300x300 pixel PNG
.RE

.SS DEFAULT OUTPUT SIZE
If you do not specify
.B --width
or
.B --height
options for the output size,
.BR rsvg-convert
will figure out a "natural size" for the SVG as follows:
.IP \(bu 2
.B SVG with width and height in pixel units (px):
.B <svg\ width="96px"\ height="192px">
For PNG output, those same dimensions in pixels are used.  For PDF/PS/EPS, that pixel size is
converted to physical units based on the DPI value (see the
.B --dpi-x
and
.B --dpi-y
options),
.IP \(bu 2
.B SVG with width and height in physical units:
.B <svg\ width="1in"\ height="2in">
For PNG output, the
.B width
and
.B height
attributes get converted to pixels, based on the DPI value (see the
.B --dpi-x
and
.B --dpi-y

NEWS
==============
Version 2.52.2
==============

Bugfixes and new features!

## New features

Thanks to Michael Howell, rsvg-convert now supports generating
multi-page PDFs in a sensible way.

With one SVG document per page, each page with the
SVG's natural size:

  rsvg-convert --format=pdf -o out.pdf a.svg b.svg c.svg

With all pages sized as portrait US Letter, and each SVG scaled to fit
so that there is a 1in margin around each page:

  rsvg-convert --format=pdf -o out.pdf \
    --page-width=8.5in --page-height=11in \
    --width=6.5in --height=8.5in --keep-aspect-ratio \
    --top=1in --left=1in \
    a.svg b.svg c.svg

Please see the man page for details.

- #738 - Support <a> elements inside <text>.  Also, support the CSS :link
  pseudo-class for matching against links. (Michael Howell)

- #649 - Support the CSS :lang() pseudo-class for matching against an
  element's xml:lang attribute. (Michael Howell)

- #790 - Support the mask-type property from SVG2.

## Fixes

- #800 - Don't panic when a shorthand property is set to
  inherit. (Michael Howell)

- #788 - Fix regression with the viewport size of interior <svg>
  elements. (Michael Howell)

- #731 - Allow length units to be case-insensitive, per SVG2. (Kolja Lampe)

## Documentation

- There is now a FEATURES.md in the repository, where you can see all
  the elements, attributes, and properties that librsvg supports.  We
  will be adding detail to this gradually.

- For developers, there is now devel-docs/adding-a-property.md with a
  tutorial on how to add support for new CSS properties.

==============
Version 2.52.1
==============

This is a bugfix release; there are no new features this time.

## Changes:

- #791 - Fix ordering of tspan inside text elements for right-to-left
  languages.

- #789 - Fix text-anchor positioning for right-to-left languages.

- #797 - Fix regression in computing sizes when an SVG has only one of
  width/height and a viewBox.  Thanks to Joshua Fogg for compiling a
  list of test cases for this.

- #565 - Spec compliance - the writing-mode property applies only to
  text elements, no to individual tspan elements.

- #794 - Fix build on big-endian platforms.

- Clarify documentation for the rsvg_handle_write() /
  rsvg_handle_close() deprecated APIs.


==============
Version 2.52.0
==============

This is a big release!  What follows is a summary from the 2.51.x
release notes; you can also read them for more detail.

The biggest user-visible change is that rsvg-convert has been ported
to Rust (Sven Neumann, Paolo Borelli), and it has new features!

## New features in rsvg-convert

### Support for physical units

rsvg-convert is now aware of physical units, and
fixes a bug where PDFs were created at the wrong size.  Do you need to
render an SVG in a PDF file, scaled to 10x10 cm, placed at a certain
position of a landscape A4 page?

  rsvg-convert --format=pdf \
    --page-width=297mm --page-height=210mm \
    --width=10cm --height=10cm --keep-aspect-ratio \
    --top=5cm --left=8cm \
    foo.svg > foo.pdf

Please see the rsvg-convert(1) man page for more details and plenty of
examples.

### Support for Accept-Language

Previously, librsvg picked up the user's language preferences through
environment variables like LANG and LC_MESSAGES.  This is inconvenient
for applications that call rsvg-convert but don't want to synthesize a
LANG variable.

There is a new option in rsvg-convert so you can pass
--accept-language=<languages> formatted as an HTTP Accept-Language
header.  This is used to specify which languages will be chosen from
elements with the "systemLanguage" attribute:

    rsvg-convert --accept-language=es-MX,en foo.svg

That command will select Mexican Spanish and English from suitable SVG
elements.  Please see the man page for details.

### Miscellaneous

rsvg-convert's default DPI is now 96, to better match W3C

rsvg-convert.c
/* -*- Mode: C; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
/* vim: set sw=4 sts=4 expandtab: */
/*

   rsvg-convert.c: Command line utility for exercising rsvg with cairo.
 
   Copyright (C) 2005 Red Hat, Inc.
   Copyright (C) 2005 Dom Lachowicz <cinamod@hotmail.com>
   Copyright (C) 2005 Caleb Moore <c.moore@student.unsw.edu.au>
   Copyright (C) 2019 Federico Mena Quintero <federico@gnome.org>
  
   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with this library; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
  
   Authors: Carl Worth <cworth@cworth.org>, 
            Caleb Moore <c.moore@student.unsw.edu.au>,
            Dom Lachowicz <cinamod@hotmail.com>,
            Federico Mena Quintero <federico@gnome.org>
*/

#define RSVG_DISABLE_DEPRECATION_WARNINGS

#include "config.h"

#include <math.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include <locale.h>
#include <glib/gi18n.h>
#include <gio/gio.h>

#ifdef G_OS_UNIX
#include <gio/gunixinputstream.h>
#endif

#ifdef G_OS_WIN32
#define WIN32_LEAN_AND_MEAN
#include <windows.h>
#include <io.h>
#include <fcntl.h>

#include <gio/gwin32inputstream.h>
#endif

#include "librsvg/rsvg-css.h"
#include "librsvg/rsvg.h"

#ifdef CAIRO_HAS_PS_SURFACE
#include <cairo-ps.h>
#endif

#ifdef CAIRO_HAS_PDF_SURFACE
#include <cairo-pdf.h>
#endif

#ifdef CAIRO_HAS_SVG_SURFACE
#include <cairo-svg.h>
#endif

#ifdef CAIRO_HAS_XML_SURFACE
#include <cairo-xml.h>
#endif

typedef enum {
    SIZE_KIND_ZOOM,
    SIZE_KIND_WH,
    SIZE_KIND_WH_MAX,
    SIZE_KIND_ZOOM_MAX
} SizeKind;

typedef struct {
    SizeKind kind;
    double x_zoom;
    double y_zoom;
    gint width;
    gint height;

    gboolean keep_aspect_ratio;
} SizeMode;

static void
get_final_size (int *width, int *height, SizeMode *real_data)
{
    double zoomx, zoomy, zoom;

    int in_width, in_height;

    in_width = *width;
    in_height = *height;

    switch (real_data->kind) {
    case SIZE_KIND_ZOOM:
        if (*width < 0 || *height < 0)
            return;

        *width = floor (real_data->x_zoom * *width + 0.5);
        *height = floor (real_data->y_zoom * *height + 0.5);
        break;

    case SIZE_KIND_ZOOM_MAX:
        if (*width < 0 || *height < 0)
            return;

        *width = floor (real_data->x_zoom * *width + 0.5);
        *height = floor (real_data->y_zoom * *height + 0.5);

        if (*width > real_data->width || *height > real_data->height) {
            zoomx = (double) real_data->width / *width;
            zoomy = (double) real_data->height / *height;
            zoom = MIN (zoomx, zoomy);

            *width = floor (zoom * *width + 0.5);
            *height = floor (zoom * *height + 0.5);
        }
        break;

    case SIZE_KIND_WH_MAX:
        if (*width < 0 || *height < 0)
            return;

        zoomx = (double) real_data->width / *width;
        zoomy = (double) real_data->height / *height;
        if (zoomx < 0)
            zoom = zoomy;
        else if (zoomy < 0)
            zoom = zoomx;
        else
            zoom = MIN (zoomx, zoomy);

        *width = floor (zoom * *width + 0.5);
        *height = floor (zoom * *height + 0.5);
        break;

    case SIZE_KIND_WH:
        if (real_data->width != -1)
            *width = real_data->width;
        if (real_data->height != -1)
            *height = real_data->height;
        break;

    default:
        g_assert_not_reached ();
    }

    if (real_data->keep_aspect_ratio) {
        int out_min = MIN (*width, *height);

        if (out_min == *width) {
            *height = in_height * ((double) *width / (double) in_width);
        } else {
            *width = in_width * ((double) *height / (double) in_height);
        }
    }
}

static void
display_error (GError * err)
{
    if (err) {
        g_printerr ("%s\n", err->message);
        g_error_free (err);
    }
}

static cairo_status_t
rsvg_cairo_write_func (void *closure, const unsigned char *data, unsigned int length)
{
    if (fwrite (data, 1, length, (FILE *) closure) == length)
        return CAIRO_STATUS_SUCCESS;
    return CAIRO_STATUS_WRITE_ERROR;
}

static char *
get_lookup_id_from_command_line (const char *lookup_id)
{
    char *export_lookup_id;

    if (lookup_id == NULL)
        export_lookup_id = NULL;
    else {
        /* rsvg_handle_has_sub() expects ids to have a '#' prepended to them, so
         * it can lookup ids in externs like "subfile.svg#subid".  For the
         * user's convenience, we include this '#' automatically; we only
         * support specifying ids from the toplevel, and don't expect users to
         * lookup things in externs.
         */
        export_lookup_id = g_strdup_printf ("#%s", lookup_id);
    }

    return export_lookup_id;
}
 
int
main (int argc, char **argv)
{
    GOptionContext *g_option_context;
    double x_zoom = 1.0;
    double y_zoom = 1.0;
    double zoom = 1.0;
    double dpi_x = 90.0;
    double dpi_y = 90.0;
    int width = -1;
    int height = -1;
    int bVersion = 0;
    char *format = NULL;
    char *output = NULL;
    char *stylesheet = NULL;
    char *export_id = NULL;
    int keep_aspect_ratio = FALSE;
    guint32 background_color = 0;
    char *background_color_str = NULL;
    gboolean using_stdin = FALSE;
    gboolean unlimited = FALSE;
    gboolean keep_image_data = FALSE;
    gboolean no_keep_image_data = FALSE;
    GError *error = NULL;

    gboolean success = TRUE;

    int i;
    char **args = NULL;
    gint n_args = 0;
    RsvgHandle *rsvg = NULL;
    cairo_surface_t *surface = NULL;
    cairo_t *cr = NULL;
    RsvgHandleFlags flags = RSVG_HANDLE_FLAGS_NONE;
    RsvgDimensionData dimensions;
    FILE *output_file = stdout;
    char *export_lookup_id;
    double unscaled_width, unscaled_height;
    int scaled_width, scaled_height;

    char *stylesheet_data = NULL;
    gsize stylesheet_data_len = 0;

    char buffer[25];
    char *endptr;
    char *source_date_epoch;
    time_t now;
    struct tm *build_time;
    unsigned long long epoch;

#ifdef G_OS_WIN32
    HANDLE handle;
#endif

    GOptionEntry options_table[] = {
        {"dpi-x", 'd', 0, G_OPTION_ARG_DOUBLE, &dpi_x,
         N_("pixels per inch [optional; defaults to 90dpi]"), N_("<float>")},
        {"dpi-y", 'p', 0, G_OPTION_ARG_DOUBLE, &dpi_y,
         N_("pixels per inch [optional; defaults to 90dpi]"), N_("<float>")},
        {"x-zoom", 'x', 0, G_OPTION_ARG_DOUBLE, &x_zoom,
         N_("x zoom factor [optional; defaults to 1.0]"), N_("<float>")},
        {"y-zoom", 'y', 0, G_OPTION_ARG_DOUBLE, &y_zoom,
         N_("y zoom factor [optional; defaults to 1.0]"), N_("<float>")},
        {"zoom", 'z', 0, G_OPTION_ARG_DOUBLE, &zoom, N_("zoom factor [optional; defaults to 1.0]"),
         N_("<float>")},
        {"width", 'w', 0, G_OPTION_ARG_INT, &width,
         N_("width [optional; defaults to the SVG's width]"), N_("<int>")},
        {"height", 'h', 0, G_OPTION_ARG_INT, &height,
         N_("height [optional; defaults to the SVG's height]"), N_("<int>")},
        {"format", 'f', 0, G_OPTION_ARG_STRING, &format,
         N_("save format [optional; defaults to 'png']"), N_("[png, pdf, ps, eps, svg, xml, recording]")},
        {"output", 'o', 0, G_OPTION_ARG_STRING, &output,
         N_("output filename [optional; defaults to stdout]"), NULL},
        {"export-id", 'i', 0, G_OPTION_ARG_STRING, &export_id,
         N_("SVG id of object to export [optional; defaults to exporting all objects]"), N_("<object id>")},
        {"keep-aspect-ratio", 'a', 0, G_OPTION_ARG_NONE, &keep_aspect_ratio,
         N_("whether to preserve the aspect ratio [optional; defaults to FALSE]"), NULL},
        {"background-color", 'b', 0, G_OPTION_ARG_STRING, &background_color_str,
         N_("set the background color [optional; defaults to None]"), N_("[black, white, #abccee, #aaa...]")},
        {"stylesheet", 's', 0, G_OPTION_ARG_FILENAME, &stylesheet, N_("Filename of CSS stylesheet"), NULL},
        {"unlimited", 'u', 0, G_OPTION_ARG_NONE, &unlimited, N_("Allow huge SVG files"), NULL},
        {"keep-image-data", 0, 0, G_OPTION_ARG_NONE, &keep_image_data, N_("Keep image data"), NULL},
        {"no-keep-image-data", 0, 0, G_OPTION_ARG_NONE, &no_keep_image_data, N_("Don't keep image data"), NULL},
        {"version", 'v', 0, G_OPTION_ARG_NONE, &bVersion, N_("show version information"), NULL},
        {G_OPTION_REMAINING, 0, 0, G_OPTION_ARG_FILENAME_ARRAY, &args, NULL, N_("[FILE...]")},
        {NULL}
    };

    /* Set the locale so that UTF-8 filenames work */
    setlocale(LC_ALL, "");

    g_option_context = g_option_context_new (_("- SVG Converter"));
    g_option_context_add_main_entries (g_option_context, options_table, NULL);
    g_option_context_set_help_enabled (g_option_context, TRUE);
    if (!g_option_context_parse (g_option_context, &argc, &argv, &error)) {
        g_option_context_free (g_option_context);
        display_error (error);
        exit (1);
    }

    g_option_context_free (g_option_context);

    if (bVersion != 0) {
        printf (_("rsvg-convert version %s\n"), VERSION);
        return 0;
    }

    if (stylesheet != NULL) {
        error = NULL;
        if (!g_file_get_contents (stylesheet, &stylesheet_data, &stylesheet_data_len, &error)) {
            g_printerr (_("Error reading stylesheet: %s\n"), error->message);
            exit (1);
        }
    }

    if (output != NULL) {
        output_file = fopen (output, "wb");
        if (!output_file) {
            g_printerr (_("Error saving to file: %s\n"), output);
            g_free (output);
            exit (1);
        }

        g_free (output);
    }
#ifdef G_OS_WIN32
    else {
        setmode (fileno (stdout), O_BINARY);
    }
#endif   

    if (args)
        while (args[n_args] != NULL)
            n_args++;

    if (n_args == 0) {
        const gchar * const stdin_args[] = { "stdin", NULL };
        n_args = 1;
        using_stdin = TRUE;
        g_strfreev (args);
        args = g_strdupv ((gchar **) stdin_args);
    } else if (n_args > 1 && (!format || !(!strcmp (format, "ps") || !strcmp (format, "eps") || !strcmp (format, "pdf")))) {
        g_printerr (_("Multiple SVG files are only allowed for PDF and (E)PS output.\n"));
        exit (1);
    }

    if (dpi_x <= 0.0 || dpi_y <= 0.0) {
        g_printerr (_("Invalid resolution\n"));
        exit (1);
    }

    if (format != NULL &&
        (g_str_equal (format, "ps") || g_str_equal (format, "eps") || g_str_equal (format, "pdf")) &&
        !no_keep_image_data)
        keep_image_data = TRUE;

    if (zoom != 1.0)
        x_zoom = y_zoom = zoom;

    if (unlimited)
        flags |= RSVG_HANDLE_FLAG_UNLIMITED;

    if (keep_image_data)
        flags |= RSVG_HANDLE_FLAG_KEEP_IMAGE_DATA;

    for (i = 0; i < n_args && success; i++) {
        GFile *file;
        GInputStream *stream;

        if (using_stdin) {
            file = NULL;
#ifdef _WIN32
            handle = GetStdHandle (STD_INPUT_HANDLE);

            if (handle == INVALID_HANDLE_VALUE) {
              gchar *emsg = g_win32_error_message (GetLastError());
              g_printerr ( _("Unable to acquire HANDLE for STDIN: %s\n"), emsg);
              g_free (emsg);
              exit (1);
            }
            stream = g_win32_input_stream_new (handle, FALSE);
#else
            stream = g_unix_input_stream_new (STDIN_FILENO, FALSE);
#endif
        } else {
            file = g_file_new_for_commandline_arg (args[i]);
            stream = (GInputStream *) g_file_read (file, NULL, &error);

            if (stream == NULL)
                goto done;
        }

        rsvg = rsvg_handle_new_from_stream_sync (stream, file, flags, NULL, &error);

    done:
        g_clear_object (&stream);
        g_clear_object (&file);

        if (error != NULL) {
            g_printerr (_("Error reading SVG:"));
            display_error (error);
            g_printerr ("\n");
            exit (1);
        }

        g_assert (rsvg != NULL);

        if (stylesheet_data != NULL) {
            if (!rsvg_handle_set_stylesheet (rsvg, stylesheet_data, stylesheet_data_len, &error)) {
                g_printerr (_("Error in stylesheet: %s\n"), error->message);
                exit (1);
            }
        }

        rsvg_handle_set_dpi_x_y (rsvg, dpi_x, dpi_y);

        export_lookup_id = get_lookup_id_from_command_line (export_id);
        if (export_lookup_id != NULL
            && !rsvg_handle_has_sub (rsvg, export_lookup_id)) {
            g_printerr (_("File %s does not have an object with id \"%s\"\n"), args[i], export_id);
            exit (1);
        }

        if (i == 0) {
            SizeMode size_data;

            if (!rsvg_handle_get_dimensions_sub (rsvg, &dimensions, export_lookup_id)) {
                g_printerr ("Could not get dimensions for file %s\n", args[i]);
                exit (1);
            }

            if (dimensions.width == 0 || dimensions.height == 0) {
                g_printerr ("The SVG %s has no dimensions\n", args[i]);
                exit (1);
            }

            unscaled_width = dimensions.width;
            unscaled_height = dimensions.height;

            /* if both are unspecified, assume user wants to zoom the image in at least 1 dimension */
            if (width == -1 && height == -1) {
                size_data.kind = SIZE_KIND_ZOOM;
                size_data.x_zoom = x_zoom;
                size_data.y_zoom = y_zoom;
                size_data.keep_aspect_ratio = keep_aspect_ratio;
            } else if (x_zoom == 1.0 && y_zoom == 1.0) {
                /* if one parameter is unspecified, assume user wants to keep the aspect ratio */
                if (width == -1 || height == -1) {
                    size_data.kind = SIZE_KIND_WH_MAX;
                    size_data.width = width;
                    size_data.height = height;
                    size_data.keep_aspect_ratio = keep_aspect_ratio;
                } else {
                    size_data.kind = SIZE_KIND_WH;
                    size_data.width = width;
                    size_data.height = height;
                    size_data.keep_aspect_ratio = keep_aspect_ratio;
                }
            } else {
                /* assume the user wants to zoom the image, but cap the maximum size */
                size_data.kind = SIZE_KIND_ZOOM_MAX;
                size_data.x_zoom = x_zoom;
                size_data.y_zoom = y_zoom;
                size_data.width = width;
                size_data.height = height;
                size_data.keep_aspect_ratio = keep_aspect_ratio;
            }

            scaled_width = dimensions.width;
            scaled_height = dimensions.height;
            get_final_size (&scaled_width, &scaled_height, &size_data);

            if (scaled_width > 32767 || scaled_height > 32767) {
                g_printerr (_("The resulting image would be larger than 32767 pixels on either dimension.\n"
                              "Librsvg currently cannot render to images bigger than that.\n"
                              "Please specify a smaller size.\n"));
                exit (1);
            }

            if (!format || !strcmp (format, "png"))
                surface = cairo_image_surface_create (CAIRO_FORMAT_ARGB32,
                                                      scaled_width, scaled_height);
#ifdef CAIRO_HAS_PDF_SURFACE
            else if (!strcmp (format, "pdf")) {
                surface = cairo_pdf_surface_create_for_stream (rsvg_cairo_write_func, output_file,
                                                               scaled_width, scaled_height);
                source_date_epoch = getenv("SOURCE_DATE_EPOCH");
                if (source_date_epoch) {
                    errno = 0;
                    epoch = strtoull(source_date_epoch, &endptr, 10);
                    if ((errno == ERANGE && (epoch == ULLONG_MAX || epoch == 0))
                            || (errno != 0 && epoch == 0)) {
                        g_printerr (_("Environment variable $SOURCE_DATE_EPOCH: strtoull: %s\n"),
                                    strerror(errno));
                        exit (1);
                    }
                    if (endptr == source_date_epoch) {
                        g_printerr (_("Environment variable $SOURCE_DATE_EPOCH: No digits were found: %s\n"),
                                    endptr);
                        exit (1);
                    }
                    if (*endptr != '\0') {
                        g_printerr (_("Environment variable $SOURCE_DATE_EPOCH: Trailing garbage: %s\n"),
                                    endptr);
                        exit (1);
                    }
                    if (epoch > ULONG_MAX) {
                        g_printerr (_("Environment variable $SOURCE_DATE_EPOCH: value must be smaller than or equal to %lu but was found to be: %llu \n"),
                                    ULONG_MAX, epoch);
                        exit (1);
                    }
                    now = (time_t) epoch;
                    build_time = gmtime(&now);
                    g_assert (strftime (buffer, sizeof (buffer), "%Y-%m-%dT%H:%M:%S%z", build_time));
                    cairo_pdf_surface_set_metadata (surface,
                                                    CAIRO_PDF_METADATA_CREATE_DATE,
                                                    buffer);
                }
            }
#endif
#ifdef CAIRO_HAS_PS_SURFACE
            else if (!strcmp (format, "ps") || !strcmp (format, "eps")){
                surface = cairo_ps_surface_create_for_stream (rsvg_cairo_write_func, output_file,
                                                              scaled_width, scaled_height);
                if(!strcmp (format, "eps"))
                    cairo_ps_surface_set_eps(surface, TRUE);
            }
#endif
#ifdef CAIRO_HAS_SVG_SURFACE
            else if (!strcmp (format, "svg")) {
                surface = cairo_svg_surface_create_for_stream (rsvg_cairo_write_func, output_file,
                                                               scaled_width, scaled_height);
                cairo_svg_surface_set_document_unit(surface, CAIRO_SVG_UNIT_PX);
            }
#endif
#ifdef CAIRO_HAS_XML_SURFACE
            else if (!strcmp (format, "xml")) {
                cairo_device_t *device = cairo_xml_create_for_stream (rsvg_cairo_write_func, output_file);
                surface = cairo_xml_surface_create (device, CAIRO_CONTENT_COLOR_ALPHA,
                                                    scaled_width, scaled_height);
                cairo_device_destroy (device);
            }
#if CAIRO_VERSION >= CAIRO_VERSION_ENCODE (1, 10, 0)
            else if (!strcmp (format, "recording"))
                surface = cairo_recording_surface_create (CAIRO_CONTENT_COLOR_ALPHA, NULL);
#endif
#endif
            else {
                g_printerr (_("Unknown output format.\n"));
                exit (1);
            }

            cr = cairo_create (surface);
            cairo_scale (cr,
                         scaled_width / unscaled_width,
                         scaled_height / unscaled_height);
        }

        // Set background color
        if (background_color_str && g_ascii_strcasecmp(background_color_str, "none") != 0) {
            RsvgCssColorSpec spec;

            spec = rsvg_css_parse_color (background_color_str);
            if (spec.kind == RSVG_CSS_COLOR_SPEC_ARGB) {
                background_color = spec.argb;
            } else {
                g_printerr (_("Invalid color specification.\n"));
                exit (1);
            }

            cairo_set_source_rgba (
                cr, 
                ((background_color >> 16) & 0xff) / 255.0, 
                ((background_color >> 8) & 0xff) / 255.0, 
                ((background_color >> 0) & 0xff) / 255.0,
                ((background_color >> 24) & 0xff) / 255.0);
            cairo_rectangle (cr, 0, 0, unscaled_width, unscaled_height);
            cairo_fill (cr);
        }

        if (export_lookup_id) {
            RsvgPositionData pos;

            if (!rsvg_handle_get_position_sub (rsvg, &pos, export_lookup_id)) {
                g_printerr (_("File %s does not have an object with id \"%s\"\n"), args[i], export_id);
                exit (1);
            }

            /* Move the whole thing to 0, 0 so the object to export is at the origin */
            cairo_translate (cr, -pos.x, -pos.y);
        }

        if (!rsvg_handle_render_cairo_sub (rsvg, cr, export_lookup_id)) {
            g_printerr ("Could not render file %s\n", args[i]);
            exit (1);
        }

        g_free (export_lookup_id);

        if (!format || !strcmp (format, "png"))
            cairo_surface_write_to_png_stream (surface, rsvg_cairo_write_func, output_file);
#if CAIRO_HAS_XML_SURFACE && CAIRO_VERSION >= CAIRO_VERSION_ENCODE (1, 10, 0)
        else if (!strcmp (format, "recording")) {
            cairo_device_t *device = cairo_xml_create_for_stream (rsvg_cairo_write_func, output_file);
            cairo_xml_for_recording_surface (device, surface);
            cairo_device_destroy (device);
        }
#endif
        else if (!strcmp (format, "xml"))
          ;
        else if (!strcmp (format, "svg") || !strcmp (format, "pdf") || !strcmp (format, "ps") || !strcmp (format, "eps"))
            cairo_show_page (cr);
        else
          g_assert_not_reached ();

        g_object_unref (rsvg);
    }

    cairo_destroy (cr);

    cairo_surface_destroy (surface);

    fclose (output_file);

    g_strfreev (args);

    return 0;
}
